### ISSUE TITLE
Crash in Kernel.ParallelCompiler

### ISSUE THREAD

**Issue by lukaszsamson** [elixir-lang/elixir#12228]
### Elixir and Erlang/OTP versions

Erlang/OTP 25 [erts-13.1.2] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [jit:ns]
Elixir 1.13.4 (compiled with Erlang/OTP 25)

### Operating system

Ubuntu 20.04.5

### Current behavior

This crash has been reported in elioxir-LS test suite on 1.13. I think it's still relevant as the code is the same on 1.14 branch.

```
GenServer ElixirLS.LanguageServer.Server terminating\n** (MatchError) no match of right hand side value: nil\n    (elixir 1.13.4) lib/kernel/parallel_compiler.ex:705: anonymous fn/3 in Kernel.ParallelCompiler.handle_deadlock/2\n    (elixir 1.13.4) lib/enum.ex:2396: Enum.\"-reduce/3-lists^foldl/2-0-\"/3\n    (elixir 1.13.4) lib/kernel/parallel_compiler.ex:701: Kernel.ParallelCompiler.handle_deadlock/2\n    (elixir 1.13.4) lib/kernel/parallel_compiler.ex:433: Kernel.ParallelCompiler.spawn_workers/7\n    (elixir 1.13.4) lib/kernel/parallel_compiler.ex:204: Kernel.ParallelCompiler.spawn_workers/5\n    (elixir 1.13.4) lib/kernel/parallel_compiler.ex:177: Kernel.ParallelCompiler.spawn_workers/3\n    (language_server 0.11.0) lib/language_server/build.ex:92: ElixirLS.LanguageServer.Build.reload_project/0\n    (language_server 0.11.0) lib/language_server/server.ex:1318: ElixirLS.LanguageServer.Server.maybe_rebuild/1\n    (language_server 0.11.0) lib/language_server/server.ex:1130: ElixirLS.LanguageServer.Server.set_settings/2\n    (language_server 0.11.0) lib/language_server/server.ex:187: ElixirLS.LanguageServer.Server.handle_cast/2\n    (stdlib 4.1.1) gen_server.erl:1123: :gen_server.try_dispatch/4\n    (stdlib 4.1.1) gen_server.erl:1200: :gen_server.handle_msg/6\n    (stdlib 4.1.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```

The code calling ParallelCompiler is rather simple
```
Kernel.ParallelCompiler.compile([mixfile])
```
It's running under `:global.trans` so there should not be a deadlock.

### Expected behavior

No crash

**Comment by josevalim** [comment]
The deadlock is within the parallel compiler, not external to it. :)

What are the contents of the `mixfile` so we can reproduce it?

**Comment by michallepicki** [comment]
There is a failure in the [elixir-ls repo](https://github.com/elixir-lsp/elixir-ls/actions/runs/3368325909/jobs/5586741832) test `test/server_test.exs:1222`

**Comment by lukaszsamson** [comment]
> What are the contents of the mixfile so we can reproduce it?

Nothing special

```
defmodule TestCodeLens.MixProject do
  use Mix.Project

  def project do
    [app: :test_code_lens, version: "0.1.0"]
  end

  def application, do: []
end
```

BTW I just hit it on local testing with 1.14

**Comment by lukaszsamson** [comment]
`handle_deadlock` gets called with
```
waiting: [],
files: [
  %{
    compiling: 0,
    file: "/Users/lukaszsamson/vscode-elixir-ls/elixir-ls/apps/elixir_ls_utils/.tmp/ElixirLS.LanguageServer.ServerTest/test returns code lenses for runnable tests with custom test paths and test pattern in umbrella apps/mix.exs",
    pid: #PID<0.520.0>,
    ref: #Reference<0.1359403991.450625537.77413>,
    timestamp: -576460747968081119,
    waiting: 0,
    warned: false
  }
]

**Comment by lukaszsamson** [comment]
Shouldn't guard https://github.com/elixir-lang/elixir/blob/a285e5876b6f69e7c340d04851aae63c8a32bf1d/lib/elixir/lib/kernel/parallel_compiler.ex#L406 also check `waiting != []`?

**Comment by lukaszsamson** [comment]
The bug is reproducible on ~20% of runs `mix test test/server_test.exs` in language_server app on elixir-ls master.
The tests that fail do 2 builds in short succesion. Mixfile module is purged and deleted before compilation. Sometimes purge returns false, sometimes there's a warning `warning: redefining module...` but those are not correlated with the crash.

**Comment by josevalim** [comment]
There is something else happening here.

You were not supposed to have more files than waiting. So something is cleaning up the waiting but not the files.

**Comment by josevalim** [comment]
OK i have pushed f9e39ef6ba5f6a9e13db67a49a86c5b819586142 which is a bit more assertive that files must be empty if spawned/waiting are empty. Because before things could have been silently working.

Other than that, I could not reproduce the failure on my machine, neither on 1.13, 1.14 or main. :(

**Comment by lukaszsamson** [comment]
@josevalim I couldn't reproduce it on master with and without your commit but I bisected the fix to https://github.com/elixir-lang/elixir/commit/6e0c721395deecfb57ebaef672d297a7cd89f271. Looking at the error raport in https://github.com/elixir-lang/elixir/issues/12087 this is likely the same problem. Can we backport it to 1.14? I don't know why it wasn't included in 1.14.1

**Comment by josevalim** [comment]
Unfortunately I am not very comfortable with backporting it. Typically changes to the compiler do not go in patch versions, unless strictly required, because of the potential impact they might have.

Thanks for verifying though! :heart:
